state-chain/chains/src/sol/api.rs

```diff
@@ -201,94 +228,80 @@ impl<Environment: SolanaEnvironment> SolanaApi<Environment> {
-					SolanaEnvAccountLookupKey::VaultProgramDataAccount,
-				)?;
-				let token_vault_pda_account =
-					Environment::lookup_account(SolanaEnvAccountLookupKey::TokenVaultPdaAccount)?;
-				let token_mint_pubkey =
-					Environment::lookup_account(SolanaEnvAccountLookupKey::TokenMintPubkey)?;
-				let token_program_id = SolAddress::from_str(TOKEN_PROGRAM_ID)
-					.expect("Preset Solana Token program ID must be valid.");
-				let token_vault_ata = Environment::lookup_account(
-					SolanaEnvAccountLookupKey::TokenVaultAssociatedTokenAccount,
-				)?;
-				let system_program_id = SolAddress::from_str(SYSTEM_PROGRAM_ID)
-					.expect("Preset Solana System Program ID account must be valid.");
-
-				// Build the Transfer instruction set for Token transfers.
-				token
-					.into_iter()
-					.map(|transfer_param| {
+					),
+					SolAsset::SolUsdc => {
+						let vault_program = Environment::lookup_account_retain(
+							SolanaEnvAccountLookupKey::VaultProgram,
+							&mut sol_environment,
+						)?;
+						let vault_program_data_account = Environment::lookup_account_retain(
+							SolanaEnvAccountLookupKey::VaultProgramDataAccount,
+							&mut sol_environment,
+						)?;
+						let sol_usdc_environment = Environment::get_token_environment(
+							&mut token_environments,
+							SolAsset::SolUsdc,
+						)?;
+						let token_program_id = SolAddress::from_str(TOKEN_PROGRAM_ID)
+							.expect("Preset Solana Token program ID must be valid.");
+						let system_program_id = SolAddress::from_str(SYSTEM_PROGRAM_ID)
+							.expect("Preset Solana System Program ID account must be valid.");
+
 						let ata =
 						crate::sol::sol_tx_core::address_derivation::derive_associated_token_account(
 							transfer_param.to,
-							token_mint_pubkey,
+							sol_usdc_environment.token_mint_pubkey,
 						)
 						.map_err(SolanaTransactionBuildingError::FailedToDeriveAddress)?;
-						Ok(SolanaInstructionBuilder::transfer_usdc_token(
+						SolanaInstructionBuilder::transfer_token(
 							ata.address,
 							transfer_param.amount,
 							transfer_param.to,
 							vault_program,
 							vault_program_data_account,
-							token_vault_pda_account,
-							token_vault_ata,
-							token_mint_pubkey,
+							sol_usdc_environment.token_vault_pda_account,
+							sol_usdc_environment.token_vault_ata,
+							sol_usdc_environment.token_mint_pubkey,
 							token_program_id,
 							system_program_id,
 							agg_key,
 							nonce_account,
 							compute_price,
-						))
-					})
-					.collect::<Result<Vec<_>, _>>()?
-			},
-		]
-		.into_iter()
-		.flatten()
-		.map(|instruction_set| Self {
-			call_type: SolanaTransactionType::Transfer,
-			transaction: SolTransaction::new_unsigned(SolMessage::new_with_blockhash(
-				&instruction_set,
-				Some(&agg_key.into()),
-				&durable_nonce.into(),
-			)),
-			_phantom: Default::default(),
-		})
-		.collect::<Vec<_>>())
+							SOL_USDC_DECIMAL,
```

Comment:
And should this be stored in the token environment lookup too, instead of hard-coding it for USDC?

Reply:
I dont think it makes sense to store it in the storage since this is a constant.

Reply:
It could be different for different tokens so we would need to have a different value for different tokens. 

See the latest commit i pushed where i made the token arms generic, the SOL_USDC_DECIMAL is now retrieved via a function that can return other constants depending on the token.